### PR TITLE
Revenge of Uplink Discount System

### PR DESCRIPTION
**Pull request by Incoming5643**
Every uplink will have three random items on sale for between 25% and 75% off (strongly weighted towards the smaller deals). 

<b>All these items are single purchase only.</b> 

Which items are on sale are unique per uplink, so don't worry too much about the world exploding to death through cheap syndibombs. 

Any item that already implies a discount is barred from being put on sale. Nuke ops get this system too, but discounts are lessened if they hit high cost items like a mech.

Son of #14160

Lowering the cost of items can help tempt antags into trying new and novel approaches to antagery. It also adds a special temptation to stealing other antags uplinks, as which items they have a discount for will be different from their own.

_Part of the unfinished business 2016 series_

:cl:
add: There's a new category in uplinks for discounted gear. These special discounts however can only be taken once, so even if you are lucky enough to see syndibombs for 75% off you won't be able to nuke the entire station with them.
/:cl:

### CODE REVIEW

**Comment by jughu** [comment]
When you get a shielded hardsuit with 75% off does this go for every uplink of the nuke ops or just all together?

**Comment by Mervill** [comment]
Anything that will encourage antags to deploy a wider variety of their arsenal is fantastic!!

**Comment by jughu** [comment]
Blacklist syndie bundle and surplus crate tho

**Comment by RandomMarine** [comment]
Helps curb metagaming, too. Less "Oh this guy has a dual esword so he can't have a freedom implant too.".

**Comment by XDTM** [comment]
I'd also appreciate random items having a higher cost/being unavailable randomly.

**Comment by Militaires** [comment]
have it so sometimes a random nuke op arsenal item is on sale for tators too, but for a very high amount of TCs

**Comment by Incoming5643** [comment]
Nuke ops each spawn with their own uplink so they each get separate discounts. By carefully shifting points around between uplinks nuke ops could probably score a few really good bargains.

Keep in mind however that there's a lot more available items in a nuke op uplink (many of which are things like ammo which doesn't exactly help without a secondary full price purchase) so the odds of someone finding a 33% off dark gygax is pretty low.

File: code/modules/uplink/uplink_item.dm [hunk 1]
```diff
@@ -30,7 +32,28 @@ var/list/uplink_items = list() // Global list so we only initialize this once.
 			if(!filtered_uplink_items[category])
 				filtered_uplink_items[category] = list()
 			filtered_uplink_items[category][item] = I
-
+			if(I.limited_stock < 0 && !I.cant_discount && I.item && I.cost > 1)
+				sale_items += I
+
+	for(var/i = 0, i < 3, i++)
```
Comment (RemieRichards) [inline code/modules/uplink/uplink_item.dm]: ```for(var/i in 1 to 3)``` (same as 0 < 3, a total of 3 steps)

File: code/modules/uplink/uplink.dm [hunk 2]
```diff
@@ -18,10 +18,12 @@ var/global/list/uplinks = list()
 	var/datum/game_mode/gamemode = null
 	var/spent_telecrystals = 0
 	var/purchase_log = ""
+	var/list/uplink_items = list()
```
Comment (RemieRichards) [inline code/modules/uplink/uplink.dm]: don't init this when in New() it's immediately being replaced

**Comment by kilkun** [comment]
while i do like the idea of reducing the viability of metagame predictions based on antag equipment, traitor and ops are in a **really** good place balance wise and im terrified that this is going to fuck that up.

**Comment by XDTM** [comment]
I'm not really sure there can be a balance, since the uplink is static while the amount of sec/robustness of crew is very variable between rounds.

**Comment by Incoming5643** [comment]
@kilkun A great number of traitor items are just "balanced" only because they're never taken. The metagame version of the uplink IS well balanced, but it's also very small.

This is totally a shake up, but it's a shake up I think stands only to benefit people in the long run.

Also if anything's too wacky to be discounted there IS a var for that down the road.

**Pull request merged**